lib/fake_async.dart

```diff
@@ -211,39 +211,58 @@ class FakeAsync {
       {Duration timeout = const Duration(hours: 1),
       bool flushPeriodicTimers = true}) {
     final absoluteTimeout = _elapsed + timeout;
-    _fireTimersWhile((timer) {
-      if (timer._nextCall > absoluteTimeout) {
-        // TODO(nweiz): Make this a [TimeoutException].
-        throw StateError('Exceeded timeout $timeout while flushing timers');
+    while (true) {
+      // With [flushPeriodicTimers] false, continue firing timers only until
+      // all remaining timers are periodic *and* every periodic timer has had
+      // a chance to run against the final value of [_elapsed].
+      if (!flushPeriodicTimers) {
+        if (_timers
+            .every((timer) => timer.isPeriodic && timer._nextCall > _elapsed)) {
+          break;
+        }
       }
 
-      if (flushPeriodicTimers) return _timers.isNotEmpty;
+      if (!runNextTimer(until: absoluteTimeout)) {
+        if (_timers.isEmpty) break;
 
-      // Continue firing timers until the only ones left are periodic *and*
-      // every periodic timer has had a change to run against the final
-      // value of [_elapsed].
-      return _timers
-          .any((timer) => !timer.isPeriodic || timer._nextCall <= _elapsed);
-    });
+        // TODO(nweiz): Make this a [TimeoutException].
+        throw StateError('Exceeded timeout $timeout while flushing timers');
+      }
+    }
   }
 
-  /// Invoke the callback for each timer until [predicate] returns `false` for
-  /// the next timer that would be fired.
+  /// Elapses time to run one timer, if any timer exists.
+  ///
+  /// Running one timer at a time, rather than just advancing time such as
+  /// with [elapse] or [flushTimers], can be useful if a test wants to run
+  /// its own logic between each timer event, for example to verify invariants
+  /// or to end the test early in case of an error.
```

Comment:
Adapted from the suggestion at https://github.com/dart-lang/fake_async/pull/85#pullrequestreview-2148110979 .

For this class of use case:
> if a test wants to simulate other asynchronous non-timer events between timer events, for example simulating UI updates or port communication.

my thinking would be that the convenient way to handle those is to create timers that trigger them. By putting them all on the same timeline measured with `Duration`, that seems like it makes for a relatively easy-to-think-about way of controlling the relative order of those events and of any timer events that are also happening.

For UI updates, in the case of Flutter with `package:flutter_test`, I believe those are already triggered by a timer.

As I see it the key things that make this API valuable, in a way that can't be substituted for by just scheduling more timers, are that the test can have its own code run after each timer (however frequent or infrequent those might be on the fake-time timeline), and that it can break out of the loop at will. The use case I have myself, described at #84 and demo'ed at https://github.com/gnprice/zulip-flutter/commit/1ad0a6f9da2b50db7102f4a82ef00046ee983a73 , is to do those things to check if an error has been recorded, and if so then to throw and stop running timers.

Reply:
Agree. If someone wants to simulate events, they still have to simulate *when* the event occurs, and then it might as well be treated as a timer event. (That also ensures that all prior due timers get handled first.)